{"nbformat":4,"nbformat_minor":0,"metadata":{"colab":{"name":"DEEPre: sequence-based enzyme EC number prediction by DL.ipynb","provenance":[],"collapsed_sections":[],"authorship_tag":"ABX9TyObZNW2D01Wxa4Lp+govAVE"},"kernelspec":{"name":"python3","display_name":"Python 3"},"language_info":{"name":"python"}},"cells":[{"cell_type":"markdown","metadata":{"id":"ejRlZ0UHuG3X"},"source":["[Repo Link](https://github.com/liyu95/Deep_learning_examples/blob/master/1.Fully_connected_psepssm_predict_enzyme/predict_enzyme.ipynb)\n","## Motivation\n","- Annotation of enzyme function: metagenomics, industrial biotech, diagnosis of enzyme deficiency-caused diseases\n","- Long time & high cost to experimentally determine the function \n","- Algo to determine enzyme function by predicting the Enzyme Commission (EC) number\n","\n","## Results\n","- end-to-end feature selection & classification model, automatic & robust feature dimensionality uniformization method\n","- instead of extracting manually crafted features from enzyme sequences, the model takes the raw sequence encoding as inputs, extracting convolutional and sequential features from the raw encoding based on the classification result to directly improve the prediction performance\n","- cross-fold validation experiments conducted on 2 large-scale datasets show that DEEPre improves the prediction performance over the previous state-of-the-art methods\n","\n","<img src = 'https://oup.silverchair-cdn.com/oup/backfile/Content_public/Journal/bioinformatics/34/5/10.1093_bioinformatics_btx680/2/m_btx680f1.jpeg?Expires=1637011796&Signature=tWmUZfXISbtufiQnpDzbtkZ3HjgbS4Pj0zi0NRhdHcdBgyjzPlIn2Xxxk2omKnYt3-OxFSo8kyn7hHpjSzOgxlt0ZlYrZ1zSxXJF9XVX47IcZGaO-55qb61QoiHEgFglXu6Jc~kNh-d38uSDzYo5pzjZcKbZaFZX2crpm6gzUS2tQczCbFRD8fikHvQDpWROXx7bMgQqxStGiZgE3VD2J2Fu4zqaxvkgg9cTLE5fGpNdS4q0lypSEeB3X6GzaCEFN9w1hng1E2tIUDvn49bJiI0Eq14RiJXMuyL15rd25OZjGWQ9EImABQErkGoh1TfJ5S0zGKL469dTbeaNJdBZPA__&Key-Pair-Id=APKAIE5G5CRDK6RD3PGA'>\n","\n","## Datasets\n","1. Ezypred: a top-down approach for predicting enzyme functional classes and subclasses. Biochem. Biophys. Res. Commun., 364, 53–59 [Shen H.B., Chou K.C., 2007](http://www.ncbi.nlm.nih.gov/pubmed/17931599). Constructed from the ENZYME database, with 40% sequence similarity cutoff (Denoted as **KNN** dataset)\n","2. Manually constructed (**NEW**) using these steps\n","  i. The SWISS-PROT (released on September 7, 2016) database was separated into enzymes and non-enzymes based on the annotation.\n","  \n","  ii. To guarantee the uniqueness and correctness, enzyme sequences with more than one set of EC numbers or incomplete EC number annotation were excluded.\n","  \n","  iii. To avoid fragment data, enzyme sequences annotated with ‘fragment’ or with <50 amino acids were excluded. Enzyme sequences with more than 5000 amino acids were also excluded.\n","  \n","  iv. To remove redundancy bias, we used CD-HIT (Fu et al., 2012) with 40% similarity threshold to sift upon the raw dataset, resulting in 22 168 low-homology enzyme sequences.\n","  \n","  v. To construct the non-enzyme part, 22 168 non-enzyme protein sequences were randomly collected from the SWISS-PROT (released on September 7, 2016) non-enzyme part, which were also subject to the (ii–iv) steps.\n","\n","3. [Benchmark dataset](http://www.ncbi.nlm.nih.gov/pubmed/22570420), referred as **COFACTOR**. Non-homologous dataset collected from PDB w/ 2 requirements: (i) the pair-wise sequence similarity within the dataset is below 30%, (ii) there is no self-BLAST hit within the dataset to ensure that there are no enzymes that are homologous to each other in this set\n","\n","All enzymes in this dataset have experimentally determined 3D structures. To avoud overlaps between the training and testing datasets, sequences contained in both our training dataset and this dataset were removed, which reduced the size of the dataset from 318 to 284\n","\n","**Background**\n","\n","According to SWISS-PROT (Bairoch and Apweiler, 2000) (released on September 7, 2016), among the 539 566 manually annotated proteins, 258 733 proteins are enzymes. Such a large number of enzymes are usually classified using the Enzyme Commission (EC) system (Cornish-Bowden, 2014), the most well-known numerical enzyme classification scheme, which specifies the function of an enzyme by four digits. This classification system has a tree structure. After the root of the tree, there are two main nodes, standing for enzyme and non-enzyme proteins, respectively. The enzyme main node extends out six successor nodes, corresponding to the six main enzyme classes: (i) oxidoreductases, (ii) transferases, (iii) hydrolases, (iv) lyases, (v) isomerases and (vi) ligases, represented by the first digit. Each main class node further extends out several subclass nodes, specifying the enzyme’s subclasses, represented by the second digit. With the same logic, the third digit indicates the enzyme’s sub-subclasses and the fourth digit denotes the sub-sub-subclasses. Take Type II restriction enzyme, which is annotated as EC 3.1.21.4, as an example, the ‘3’ denotes that it is an hydrolase; the ‘1’ indicates that it acts on ester bonds; the ‘21’ shows that it is an endodeoxyribonuclease producing 5-phosphomonoesters; and the ‘4’ suggests that it is a Type II site-specific deoxyribonuclease. By predicting the EC numbers precisely, computational methods can annotate the function of enzymes. It should also be noted that a substantial number of enzymes annotated with some reactions in databases such as UniProt or Brenda do not have EC numbers associated, which is out of the scope of this study."]},{"cell_type":"code","metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"yC2IZ9AhzqNA","executionInfo":{"status":"ok","timestamp":1641376013460,"user_tz":-480,"elapsed":2429,"user":{"displayName":"Indrik Wijaya","photoUrl":"https://lh3.googleusercontent.com/a/default-user=s64","userId":"00208067985282844199"}},"outputId":"2965c3d9-619f-429e-d1ca-bfd21357e74c"},"source":["! wget https://github.com/liyu95/Deep_learning_examples/raw/master/1.Fully_connected_psepssm_predict_enzyme/Pfam_model_names_list.pickle\n","! wget https://github.com/liyu95/Deep_learning_examples/raw/master/1.Fully_connected_psepssm_predict_enzyme/Pfam_name_list_new_data.pickle\n","! wget https://github.com/liyu95/Deep_learning_examples/raw/master/1.Fully_connected_psepssm_predict_enzyme/Pfam_name_list_non_enzyme.pickle"],"execution_count":1,"outputs":[{"output_type":"stream","name":"stdout","text":["--2022-01-05 09:46:50--  https://github.com/liyu95/Deep_learning_examples/raw/master/1.Fully_connected_psepssm_predict_enzyme/Pfam_model_names_list.pickle\n","Resolving github.com (github.com)... 140.82.114.3\n","Connecting to github.com (github.com)|140.82.114.3|:443... connected.\n","HTTP request sent, awaiting response... 302 Found\n","Location: https://raw.githubusercontent.com/liyu95/Deep_learning_examples/master/1.Fully_connected_psepssm_predict_enzyme/Pfam_model_names_list.pickle [following]\n","--2022-01-05 09:46:50--  https://raw.githubusercontent.com/liyu95/Deep_learning_examples/master/1.Fully_connected_psepssm_predict_enzyme/Pfam_model_names_list.pickle\n","Resolving raw.githubusercontent.com (raw.githubusercontent.com)... 185.199.108.133, 185.199.109.133, 185.199.110.133, ...\n","Connecting to raw.githubusercontent.com (raw.githubusercontent.com)|185.199.108.133|:443... connected.\n","HTTP request sent, awaiting response... 200 OK\n","Length: 313298 (306K) [text/plain]\n","Saving to: ‘Pfam_model_names_list.pickle’\n","\n","Pfam_model_names_li 100%[===================>] 305.96K  --.-KB/s    in 0.03s   \n","\n","2022-01-05 09:46:51 (9.28 MB/s) - ‘Pfam_model_names_list.pickle’ saved [313298/313298]\n","\n","--2022-01-05 09:46:51--  https://github.com/liyu95/Deep_learning_examples/raw/master/1.Fully_connected_psepssm_predict_enzyme/Pfam_name_list_new_data.pickle\n","Resolving github.com (github.com)... 140.82.113.3\n","Connecting to github.com (github.com)|140.82.113.3|:443... connected.\n","HTTP request sent, awaiting response... 302 Found\n","Location: https://raw.githubusercontent.com/liyu95/Deep_learning_examples/master/1.Fully_connected_psepssm_predict_enzyme/Pfam_name_list_new_data.pickle [following]\n","--2022-01-05 09:46:51--  https://raw.githubusercontent.com/liyu95/Deep_learning_examples/master/1.Fully_connected_psepssm_predict_enzyme/Pfam_name_list_new_data.pickle\n","Resolving raw.githubusercontent.com (raw.githubusercontent.com)... 185.199.108.133, 185.199.109.133, 185.199.110.133, ...\n","Connecting to raw.githubusercontent.com (raw.githubusercontent.com)|185.199.108.133|:443... connected.\n","HTTP request sent, awaiting response... 200 OK\n","Length: 1611684 (1.5M) [text/plain]\n","Saving to: ‘Pfam_name_list_new_data.pickle’\n","\n","Pfam_name_list_new_ 100%[===================>]   1.54M  --.-KB/s    in 0.05s   \n","\n","2022-01-05 09:46:52 (28.9 MB/s) - ‘Pfam_name_list_new_data.pickle’ saved [1611684/1611684]\n","\n","--2022-01-05 09:46:52--  https://github.com/liyu95/Deep_learning_examples/raw/master/1.Fully_connected_psepssm_predict_enzyme/Pfam_name_list_non_enzyme.pickle\n","Resolving github.com (github.com)... 140.82.113.4\n","Connecting to github.com (github.com)|140.82.113.4|:443... connected.\n","HTTP request sent, awaiting response... 302 Found\n","Location: https://raw.githubusercontent.com/liyu95/Deep_learning_examples/master/1.Fully_connected_psepssm_predict_enzyme/Pfam_name_list_non_enzyme.pickle [following]\n","--2022-01-05 09:46:52--  https://raw.githubusercontent.com/liyu95/Deep_learning_examples/master/1.Fully_connected_psepssm_predict_enzyme/Pfam_name_list_non_enzyme.pickle\n","Resolving raw.githubusercontent.com (raw.githubusercontent.com)... 185.199.108.133, 185.199.109.133, 185.199.110.133, ...\n","Connecting to raw.githubusercontent.com (raw.githubusercontent.com)|185.199.108.133|:443... connected.\n","HTTP request sent, awaiting response... 200 OK\n","Length: 1113123 (1.1M) [text/plain]\n","Saving to: ‘Pfam_name_list_non_enzyme.pickle’\n","\n","Pfam_name_list_non_ 100%[===================>]   1.06M  --.-KB/s    in 0.05s   \n","\n","2022-01-05 09:46:52 (21.3 MB/s) - ‘Pfam_name_list_non_enzyme.pickle’ saved [1113123/1113123]\n","\n"]}]},{"cell_type":"markdown","metadata":{"id":"3I5Jn2euzxfD"},"source":["### Load related packages\n"]},{"cell_type":"code","metadata":{"id":"zXVcWuaS0HlX","executionInfo":{"status":"ok","timestamp":1641376016362,"user_tz":-480,"elapsed":2907,"user":{"displayName":"Indrik Wijaya","photoUrl":"https://lh3.googleusercontent.com/a/default-user=s64","userId":"00208067985282844199"}}},"source":["import pickle\n","import numpy as np\n","import keras\n","import tensorflow as tf\n","from sklearn.model_selection import train_test_split\n","from keras.datasets import mnist\n","from keras.models import Sequential\n","from keras.layers import Dense, Dropout\n","from tensorflow.keras.optimizers import RMSprop"],"execution_count":2,"outputs":[]},{"cell_type":"code","metadata":{"id":"eA94VxFU0Sh-","executionInfo":{"status":"ok","timestamp":1641376017842,"user_tz":-480,"elapsed":497,"user":{"displayName":"Indrik Wijaya","photoUrl":"https://lh3.googleusercontent.com/a/default-user=s64","userId":"00208067985282844199"}}},"source":["## Helper function to load data\n","def Pfam_from_file_encoding(name_list_pickle_filename, model_names_list_filename):\n","  with open(name_list_pickle_filename, 'rb') as f:\n","    name_list = pickle.load(f)\n","  \n","  with open(model_names_list_filename, 'rb') as f:\n","    model_list = pickle.load(f)\n","  \n","  encoding = []\n","\n","  for i in range(len(name_list)):\n","    if i % 10000 == 0:\n","      print('Processing %dth sequence.'%i)\n","    single_encoding = np.zeros(16306)\n","\n","    if name_list[i] != []:\n","      for single_name in name_list[i]:\n","        single_encoding[model_list.index(single_name)] = 1\n","    encoding.append(single_encoding)\n","  \n","  return encoding"],"execution_count":3,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"54BR-gOZ0ctA"},"source":["### Load the data"]},{"cell_type":"code","metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"aTdzIMsJ1S4z","executionInfo":{"status":"ok","timestamp":1641376035227,"user_tz":-480,"elapsed":17391,"user":{"displayName":"Indrik Wijaya","photoUrl":"https://lh3.googleusercontent.com/a/default-user=s64","userId":"00208067985282844199"}},"outputId":"3eb0048d-3f18-43d9-8f33-55848c360f59"},"source":["enzyme_feature = Pfam_from_file_encoding('Pfam_name_list_new_data.pickle',\n","                                         'Pfam_model_names_list.pickle')\n","\n","non_enzyme_feature = Pfam_from_file_encoding('Pfam_name_list_non_enzyme.pickle',\n","                                         'Pfam_model_names_list.pickle')\n"],"execution_count":4,"outputs":[{"output_type":"stream","name":"stdout","text":["Processing 0th sequence.\n","Processing 10000th sequence.\n","Processing 20000th sequence.\n","Processing 0th sequence.\n","Processing 10000th sequence.\n","Processing 20000th sequence.\n"]}]},{"cell_type":"code","metadata":{"id":"tJhMulW11qfW"},"source":["feature = np.concatenate([enzyme_feature, non_enzyme_feature], axis=0)\n","label = np.concatenate([np.ones([22168,1]), np.zeros([22168,1])], axis=0).flatten()\n","label = tf.keras.utils.to_categorical(label,num_classes=2)"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["Session crashed because of insufficient RAM!"],"metadata":{"id":"1T-7yhyy1xWw"}},{"cell_type":"code","source":[""],"metadata":{"id":"eYXsO4nZpmMA"},"execution_count":null,"outputs":[]}]}